mapiproxy/libmapistore/mapistore_notification.c

```diff
@@ -3,525 +3,1355 @@
+
+	*uuid = r.v.v1.uuid;
+	*cnp = talloc_strdup(mem_ctx, r.v.v1.cn);
+	MAPISTORE_RETVAL_IF(!*cnp, MAPISTORE_ERR_NO_MEMORY, local_mem_ctx);
+
+	talloc_free(local_mem_ctx);
+
+	return MAPISTORE_SUCCESS;
+}
+
+
+/**
+   \details Generate the resolver key
+
+   \param mem_ctx pointer to the memory context
+   \param cn the common name to compute
+   \param _key pointer on pointer to the key to return
+
+   \note the caller is responsible for freeing _key
+
+   \return MAPISTORE_SUCCES on success, otherwise MAPISTORE error
+ */
+static enum mapistore_error mapistore_notification_resolver_set_key(TALLOC_CTX *mem_ctx,
+								    const char *cn,
+								    char **_key)
+{
+	char	*key = NULL;
+
+	/* Sanity checks */
+	MAPISTORE_RETVAL_IF(!cn, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+	MAPISTORE_RETVAL_IF(!strlen(cn), MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+	MAPISTORE_RETVAL_IF(!_key, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+
+	key = talloc_asprintf(mem_ctx, MSTORE_MEMC_FMT_RESOLVER, cn);
+	MAPISTORE_RETVAL_IF(!key, MAPISTORE_ERR_NO_MEMORY, NULL);
+
+	*_key = key;
+	return MAPISTORE_SUCCESS;
+}
+
+
+/**
+   \details Add a record to the resolver
+
+   \param mstore_ctx pointer to the mapistore context
+   \param cn the common name (key) to use for host registration
+   \param host the host to register
+
+   \note This function acts as a wrapper and manages both the creation
+   of a new key/value pair and the update of an existing record. The
+   following logic is implemented:
+
+   [START] Does the key exist?
+	If yes: Update record
+	If no: Insert key
+
+   \todo Note that a race condition is possible between the exist
+   check and the addition of the record, where the same key could have
+   been created by another instance, leading to the failure of the add
+   operation. While this scenario is very unlikely to happen in real
+   world, a retry with a counter should be implemented in further
+   iterations to reasonably circumvent this use case.
+
+   \return MAPISTORE_SUCCESS on success, otherwise MAPISTORE error
+ */
+_PUBLIC_ enum mapistore_error mapistore_notification_resolver_add(struct mapistore_context *mstore_ctx,
+								  const char *cn,
+								  const char *host)
+{
+	TALLOC_CTX				*mem_ctx;
+	enum mapistore_error			retval;
+	struct mapistore_notification_resolver	r;
+	struct ndr_push				*ndr;
+	enum ndr_err_code			ndr_err_code;
+	memcached_return			rc = MEMCACHED_ERROR;
+	char					*key = NULL;
+
+	/* Sanity checks */
+	MAPISTORE_RETVAL_IF(!mstore_ctx, MAPISTORE_ERR_NOT_INITIALIZED, NULL);
+	MAPISTORE_RETVAL_IF(!cn, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+	MAPISTORE_RETVAL_IF(!host, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+	MAPISTORE_RETVAL_IF(!mstore_ctx->notification_ctx, MAPISTORE_ERR_NOT_AVAILABLE, NULL);
+	MAPISTORE_RETVAL_IF(!mstore_ctx->notification_ctx->memc_ctx, MAPISTORE_ERR_NOT_AVAILABLE, NULL);
+
+	/* memcached does not allow key with space */
```

Comment:
Would it be better to move this check in `mapistore_notification_resolver_set_key` where we are generating the key for memcached entry?